### ISSUE TITLE
Probe mode for efficient unit and load testing and mock emulation of Transformer and Model

### ISSUE THREAD

**Issue by qualiteger** [qualiteg/ChatStream#34]
プローブモード、Transformer/Modelのモックエミュレーション実行を導入する

chat_core.py で　PROBE_MODE = True
つまりプローブモード有効にした場合

- 現在 chat_core で入出力されているデータをファイルに出力する

- 出力されたファイルはデフォルトで `[home_dir]/.cache/chatstream/probe_data/[dirname]` に保存される

- [dirname] は YYYYmmdd_hhmmss フォーマットからなる。

- このディレクトリにある tokens.json に入出力データが JSON 形式で格納される

- tokens.json は以下のような形式となる。

```
  "seed": 300,
  "tok_dict": {
    "special_tokens": {
      "bos_token": "<s>",
      "eos_token": "</s>",
      "unk_token": "[UNK]",
      "sep_token": "[SEP]",
      "pad_token": "[PAD]",
      "cls_token": "[CLS]",
      "mask_token": "[MASK]"
    },
    "bos_token_id": 2,
    "eos_token_id": 3,
    "pad_token_id": 0,
    "unk_token_id": 1,
    "bos_token": "<s>",
    "eos_token": "</s>",
    "pad_token": "[PAD]",
    "unk_token": "[UNK]"
  },
  "ids_to_logits_dict": {
    "6822_276_13856_407_4006_362_537_28166_18026_392_1615_4506_536_583_2680_1013_276_263": "logits-6822__c489e140-0914-4c77-a460-d89438de19b2.pt",
    "6822_276_13856_407_4006_362_537_28166_18026_392_1615_4506_536_583_2680_1013_276_263_1162": "logits-1162__c9c652a7-a0d6-4843-956b-f8af748a8acb.pt",
  },
  "prompt_to_ids_dict": {
    "ユーザー: タイタニックという映画を知っていますか？<NL>システム: ": {
      "ids": [
        6822,
        276,
        13856,
	●中略●
  ],
      "add_special_tokens": false,
      "type": "encode",
      "is_first": true
    }
  },
  "ids_to_token_dict": {
    "6822_276_13856_407_4006_362_537_28166_18026_392_1615_4506_536_583_2680_1013_276_263_1162": "ユーザー: タイタニックという映画を知っていますか?<NL>システム: タイ",
    "6822_276_13856_407_4006_362_537_28166_18026_392_1615_4506_536_583_2680_1013_276_263_1162_407": "ユーザー: タイタニックという映画を知っていますか?<NL>システム: タイタ",
		●中略●

},
  "ids_to_token_dict_single": {
    "6822": "ユーザー",
    "276": ":",
    "13856": "タイ",
	●中略●
   "15048": "を誇る",
    "5736": "ました",
    "3": ""
  }
}
```

ids_to_logits_dict　はModelによるlogits生成結果をトークンid列から取得する。

logits は torch.tensor なので、*.pt ファイルとして出力している

これらのファイルは MockTransformer によって読み込むことができる。

```
mock_transformer = MockTransformer(parent_dir_path="somewhere",dirname="20230828_164804")

model = mock_transformer.get_model()

tokenizer = mock_transformer.get_tokenizer()
```

これにより、実際の model,tokenizer を初期化することなく
プローブモードで入力した展開と同じように chat_prompt を構築することで同一の結果を得ることができる。

この仕組みはおもに単体テスト、負荷テストでの利用を想定している。

大きなモデルになるほど model,tokenizer の読み込みに時間かかるため本仕組みによりmodel,tokenizer そのものの検証以外で単体テストや負荷テストの効率的な実行を行うことができる。

**Comment by qualiteger** [comment]
Probeモードによる記録方法

1. chat_core.py でPROBE_MODE = Trueとし、webチャットを起動する。
2. 何ターンか会話やりとりをする。
3. [home_dir]/.cache/chatstream/probe_data/[dirname]にデータが出力される

モックモードの利用方法

1. chat_core.py でPROBE_MODE = Falseとする
2. 以下のようにモックからモデル、トークナイザを得る

```
mock_transformer = MockTransformer(dirname="20230828_164804")

model = mock_transformer.get_model()

tokenizer = mock_transformer.get_tokenizer()
```

3. Webチャット起動

データディレクトリを移動した場合は 以下のように parent_dir_path をセットする。

mock_transformer = MockTransformer(parent_dir_path="hogehoge" ,dirname="20230828_164804")
